keps/sig-node/20190604-rootless.md

```diff
@@ -0,0 +1,144 @@
+replaces:
+  - https://github.com/kubernetes/enhancements/pull/1084
+---
+
+# Rootless mode
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Implementation Details](#implementation-details)
+  - [User namespaces](#user-namespaces)
+    - [Paths](#paths)
+    - [Required changes to Kubernetes](#required-changes-to-kubernetes)
+  - [Network namespaces](#network-namespaces)
+    - [Required changes to Kubernetes](#required-changes-to-kubernetes-1)
+  - [cgroup](#cgroup)
+    - [Required changes to Kubernetes](#required-changes-to-kubernetes-2)
+- [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing](#testing)
+- [History](#history)
+<!-- /toc -->
+
+## Summary
+
+Allow running the entire Kubernetes components (`kubelet`, CRI, OCI, CNI, and all `kube-*`) as a non-root user on the host.
+
+See [FOSDEM 2019 talk "Rootless Kubernetes"](https://www.slideshare.net/AkihiroSuda/rootless-kubernetes) by [@AkihiroSuda](https://github.com/AkihiroSuda) and [@giuseppe](https://github.com/giuseppe) for the overview of Rootless mode.
+
+Rootless mode has been already adopted by k3s, though it doesn't support setting resource limitation as it lacks support for cgroup.
+
+In this KEP, the resource limitation is planned to be supported using cgroup v2, which officially supports delegating its control to a non-root user.
+
+## Motivation
+
+* Protect the host from potential container-breakout vulnerabilities
+* Allow users of shared machines (especially HPC) to run Kubernetes without the risk of accidentally breaking their colleagues' environments.
+  Not recommended for real multi-tenancy where the users cannot be trusted.
+* Safe Kubernetes-on-Kubernetes
+
+## Goals
+* Allow `kubelet` and `kube-proxy` being executed inside user namespaces create by a non-root user
+* Allow `kube-proxy` to propagate NodePort ports from its network namespace into the initial network namespace
+* Allow `kubelet` to configure cgroup v2 via a user-instance of systemd 
+
+## Implementation Details
+
+### User namespaces
+All components need to be executed inside a user namespace (along with a network namespace and a mount namespace)
+to gain fake-root privileges, mostly for network and mount operations.
+
+The user namespace is expected be created using [RootlessKit](https://github.com/rootless-containers/rootlesskit).
+RootlessKit has been already adopted by Docker, BuildKit, [Usernetes](https://github.com/rootless-containers/usernetes), and k3s.
+
+#### Paths
+
+Some paths like `/var/log/pods` are hardcoded and hard to change.
+As RootlessKit bind-mounts directories writable by the user on these paths, Kubernetes does not need to be updated.
+
+#### Required changes to Kubernetes
+`kubelet` and `kube-proxy` need to be update to ignore sysctl errors, because most of sysctl parameters cannot be modified in the user namespace.
```

Comment:
the pod spec supports sysctl via the pod security context, i guess this means that a rootless kubelet would not be able to support unsafe sysctls which is sensible since allowed-unsafe-sysctls is a kubelet option.  this does lead into a broader question though worth enumerating which is are there other elements of the existing pod spec that we could enumerate would not function in a rootless mode?

Reply:
Sorry to cause confusion, I was talking about the sysctl parameters configured by `kubelet` itself, and this sentence is unrelated to `.spec.securityContext.sysctls` in Pod manifests.

Creating Pod manifests with such sysctl parameters would fail. If this behavior is problematic, user should write a Mutating Admission Webhook to remove such sysctl parameters from the Pod manifest.